assets/src/edit-story/components/checklist/checks/videoOptimization.js

```diff
@@ -0,0 +1,151 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { __ } from '@web-stories-wp/i18n';
+import styled from 'styled-components';
+
+/**
+ * Internal dependencies
+ */
+import { useCallback, useMemo, useState } from 'react';
+import { useLocalMedia, useStory } from '../../../app';
+import { VIDEO_SIZE_THRESHOLD } from '../../../app/media/utils/useFFmpeg';
+import { PRIORITY_COPY } from '../constants';
+import { LayerThumbnail, Thumbnail, THUMBNAIL_TYPES } from '../../thumbnail';
+import {
+  CARD_TYPE,
+  ChecklistCard,
+  DefaultFooterText,
+} from '../../checklistCard';
+import { filterStoryElements } from '../utils/filterStoryElements';
+import { Button, BUTTON_SIZES, BUTTON_TYPES } from '../../../../design-system';
+import { useHighlights } from '../../../app/highlights';
+
+const OptimizeButton = styled(Button)`
+  margin-top: 4px;
+  border: ${({ theme }) => `1px solid ${theme.colors.border.defaultNormal}`};
+`;
+
+export function videoElementOptimized(element = {}) {
+  if (element.resource?.isTranscoding) {
+    return true;
+  }
+
+  if (
+    element.type !== 'video' ||
+    element.resource?.local ||
+    element.resource?.isOptimized
+  ) {
+    return false;
+  }
+
+  const videoArea =
+    (element.resource?.height ?? 0) * (element.resource?.width ?? 0);
+  const isLargeVideo =
+    videoArea >= VIDEO_SIZE_THRESHOLD.WIDTH * VIDEO_SIZE_THRESHOLD.HEIGHT;
+  return isLargeVideo;
+}
+
+export const BulkVideoOptimization = () => {
+  const story = useStory(({ state }) => state);
+  const unoptimizedVideos = filterStoryElements(story, videoElementOptimized);
+
+  const setHighlights = useHighlights(({ setHighlights }) => setHighlights);
+
+  const { optimizeVideo } = useLocalMedia((state) => ({
+    optimizeVideo: state.actions.optimizeVideo,
+  }));
+
+  const [isOptimizing, setIsOptimizing] = useState(false);
+
+  const sequencedVideoOptimization = useCallback(() => {
+    return new Promise((res) => {
+      unoptimizedVideos
+        .filter(({ resource }) => !resource?.isTranscoding)
+        .map(({ resource }) => async () => {
+          await optimizeVideo({ resource });
+        })
+        .reduce((p, fn) => p.then(fn), Promise.resolve())
+        .then(() => {
+          res();
+        });
+    });
+  }, [unoptimizedVideos, optimizeVideo]);
+
+  const handleUpdateVideos = useCallback(async () => {
+    setIsOptimizing(true);
+    await sequencedVideoOptimization();
+    setIsOptimizing(false);
+  }, [sequencedVideoOptimization]);
```

Comment:
Do we need to worry about any scenario here were this callback updates while in the midst of the:
```
await sequencedVideoOptimization();
```
promise?

Or is it fine since we're filtering out all the `isTranscoding` videos?

Reply:
I'm not sure and the state management here has been hard for me to track so I'm sorry I'm not more confident here. If there's something wrong with my reasoning below let me know.

`sequencedVideoOptimization` will update when `unoptimizedVideos` and `optimizeVideo` changes, which will happen as they get uploaded and start being transcoded (`isOptimized` is set before they start getting transcoded). 

It's a callback not an effect so it will close off to those updates at the time of pressing the button, which you can only press if nothing is being transcoded at the time.

so... i guess one thing that could happen is the callback would update if the user added a new unoptimized video that needed optimizing to their story while things are being transcoded from their original button click. and _that_ video would show as a thumbnail and would not get optimized as a result of the last button click. when the other videos finished transcoding or canceled,  is that right?

so I think I could add an effect that if things are being transcoded... and a new unoptimized video is being added to the story to trigger another `optimizeVideo` call? we're kind of making an assumption there. I think it would probably be safe to make the assumption that if they press the button they want all their videos optimized, but I'm a little hesitant to deal with this case in this PR since the state management might change. :/ Maybe I'm also missing another case.

Reply:
I think the fact that it's tied to a button click event is probably fine for it not re-firing. I guess when I was looking at it, I thought it might be getting called in an effect which would be more volatile. Yea, I'm fine with leaving it as is for right now. Was just curious and trying to learn the exact state flow here, which is tricky, but works fine.